cpu/stm32l1/periph/gpio.c

```diff
@@ -24,545 +26,196 @@
-#endif
+    else {
+        NVIC_EnableIRQ(EXTI15_10_IRQn);
     }
-
-    NVIC_EnableIRQ(gpio_irq_map[dev]);
-
-    /* set callback */
-    gpio_config[dev].cb = cb;
-    gpio_config[dev].arg = arg;
-
-    /* configure the event that triggers an interrupt */
+    /* configure the active edge(s) */
     switch (flank) {
         case GPIO_RISING:
-            EXTI->RTSR |= (1 << pin);
-            EXTI->FTSR &= ~(1 << pin);
+            EXTI->RTSR |= (1 << pin_num);
+            EXTI->FTSR &= ~(1 << pin_num);
             break;
         case GPIO_FALLING:
-            EXTI->RTSR &= ~(1 << pin);
-            EXTI->FTSR |= (1 << pin);
+            EXTI->RTSR &= ~(1 << pin_num);
+            EXTI->FTSR |= (1 << pin_num);
             break;
         case GPIO_BOTH:
-            EXTI->RTSR |= (1 << pin);
-            EXTI->FTSR |= (1 << pin);
+            EXTI->RTSR |= (1 << pin_num);
+            EXTI->FTSR |= (1 << pin_num);
             break;
     }
-
+    /* enable specific pin as exti sources */
+    SYSCFG->EXTICR[pin_num >> 2] &= ~(0xf << ((pin_num & 0x03) * 4));
+    SYSCFG->EXTICR[pin_num >> 2] |= (port_num << ((pin_num & 0x03) * 4));
     /* clear any pending requests */
-    EXTI->PR = (1 << pin);
-    /* unmask the pins interrupt channel */
-    EXTI->IMR |= (1 << pin);
-
+    EXTI->PR = (1 << pin_num);
+    /* enable interrupt for EXTI line */
+    EXTI->IMR |= (1 << pin_num);
     return 0;
 }
 
-void gpio_irq_enable(gpio_t dev)
+void gpio_init_af(gpio_t pin, gpio_af_t af)
 {
-    uint8_t pin;
+    GPIO_TypeDef *port = _port(pin);
+    uint32_t pin_num = _pin_num(pin);
 
-    if (dev >= GPIO_NUMOF) {
-        return;
-    }
-
-    pin = gpio_pin_map[dev];
-    EXTI->IMR |= (1 << pin);
+    /* set pin to AF mode */
+    port->MODER &= ~(3 << (2 * pin_num));
+    port->MODER |= (2 << (2 * pin_num));
+    /* set selected function */
+    port->AFR[(pin_num > 7) ? 1 : 0] &= ~(0xf << ((pin_num & 0x07) * 4));
+    port->AFR[(pin_num > 7) ? 1 : 0] |= (af << ((pin_num & 0x07) * 4));
 }
 
-void gpio_irq_disable(gpio_t dev)
+void gpio_irq_enable(gpio_t pin)
 {
-    uint8_t pin;
-
-    if (dev >= GPIO_NUMOF) {
-        return;
-    }
-
-    pin = gpio_pin_map[dev];
-    EXTI->IMR &= ~(1 << pin);
+    EXTI->IMR |= (1 << _pin_num(pin));
 }
 
-int gpio_read(gpio_t dev)
+void gpio_irq_disable(gpio_t pin)
 {
-    GPIO_TypeDef *port;
-    uint8_t pin;
-
-    if (dev >= GPIO_NUMOF) {
-        return -1;
-    }
+    EXTI->IMR &= ~(1 << _pin_num(pin));
+}
 
-    port = gpio_port_map[dev];
-    pin = gpio_pin_map[dev];
+int gpio_read(gpio_t pin)
+{
+    GPIO_TypeDef *port = _port(pin);
+    uint32_t pin_num = _pin_num(pin);
 
-    if (port->MODER & (1 << (pin * 2))) {       /* if configured as output */
-        return port->ODR & (1 << pin);          /* read output data register */
+    if (port->MODER & (3 << (pin_num * 2))) {   /* if configured as output */
+        return port->ODR & (1 << pin_num);      /* read output data reg */
     } else {
```

Comment:
According to our coding conventions, could you put the else branch in a new line please.